### ISSUE TITLE
Package request: Replace gnu-bc with "bsd"-bc

### ISSUE THREAD

**Issue by bugcrazy** [void-linux/void-packages#7195]
The gnu-bc has not been updated since 2017, nor does it have current Posix compliance. This project https://github.com/gavinhoward/bc is updated and complies with recent Posix standard.

**Comment by maxice8** [comment]
On Tue, Jan 15, 2019 at 04:34:57PM -0800, bugcrazy wrote:
> Yes, this is it! https://www.adelielinux.org/announcements/1.0-BETA2.html
> 
> "GNU bc has been replaced with a more POSIX-conformant implementation."
> 
> -- 
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub:
> https://github.com/void-linux/void-packages/issues/7195#issuecomment-454605823

Have they been able to easily build Linux ? Concerns were raised in IRC

**Comment by bugcrazy** [comment]
It was built easily, you can see through the link https://code.foxkit.us/adelie/packages/blob/master/system/bc/APKBUILD

**Comment by maxice8** [comment]
On Tue, Jan 15, 2019 at 04:57:01PM -0800, bugcrazy wrote:
> It was built easily, you can see through the link https://code.foxkit.us/adelie/packages/blob/master/system/bc/APKBUILD
> 
> -- 
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub:
> https://github.com/void-linux/void-packages/issues/7195#issuecomment-454610084

Building is no problem i have a template ready that install bsd-bc and bsd-dc, i mean transitioning to it as the system's main bc and dc i don't want to break building linux4.20 or any package at all for that matter.

Plus it would be nice if it packaged a test-suite, even if it's GNU bc's one

**Comment by maxice8** [comment]
A few quips that prevent us from transitioning to it.

Test suite fails on 1.0
Manuals aren't present on 1.0 and are in markdown (pandoc conversion was suboptimal)

**Comment by bugcrazy** [comment]
The tests that are the softwares https://github.com/fivepiece/gnu-bc/tree/1.07.1-fix_arrays/Test  and https://github.com/gavinhoward/bc/tree/master/tests

**Comment by bugcrazy** [comment]
> A few quips that prevent us from transitioning to it.
> 
> Test suite fails on 1.0
> Manuals aren't present on 1.0 and are in markdown (pandoc conversion was suboptimal)

Which test suite do you use?

**Comment by maxice8** [comment]
On Tue, Jan 15, 2019 at 05:48:50PM -0800, bugcrazy wrote:
> > A few quips that prevent us from transitioning to it.
> > 
> > Test suite fails on 1.0
> > Manuals aren't present on 1.0 and are in markdown (pandoc conversion was suboptimal)
> 
> Which test suite do you use?
> 
> -- 
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub:
> https://github.com/void-linux/void-packages/issues/7195#issuecomment-454620332

The one via make test_all on the 1.0 tarball

**Comment by maxice8** [comment]
```
***********************************************************************

Running bc tests...

Running bc decimal...
Running bc add...
Running bc subtract...
Running bc multiply...
Running bc divide...
Running bc modulus...
Running bc power...
Running bc sqrt...
Running bc trunc...
Running bc places...
Running bc shift...
Running bc vars...
Running bc boolean...
Running bc strings...
Running bc letters...
Generating bc parse...

Parse error: bad print statement
    <stdin>:1

make: *** [Makefile:172: test_bc] Error 1
=> ERROR: bsd-bc-1.1rc3_1: do_check: '${make_cmd} ${make_check_args} ${make_check_target}' exited with 2
=> ERROR:   in do_check() at common/build-style/configure.sh:32
```

**Comment by maxice8** [comment]
Another quip is that cross-compilation is broken, it tries to execute compiled binaries.

**Comment by awilfox** [comment]
No issues building Linux 4.14 here, but we don't ship anything close to bleeding edge (especially with the boot issues that 4.20 and 5.0-rc are having on PowerPC).

That seems to be 1.1rc3 that is failing; I'd assume that is not released yet for a reason :)  1.0 is working fine for us.

**Comment by maxice8** [comment]
On Wed, Jan 16, 2019 at 10:57:01AM -0800, A. Wilcox wrote:
> No issues building Linux 4.14 here, but we don't ship anything close to bleeding edge (especially with the boot issues that 4.20 and 5.0-rc are having on PowerPC).
> 
> That seems to be 1.1rc3 that is failing; I'd assume that is not released yet for a reason :)  1.0 is working fine for us.
> 
> -- 
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub:
> https://github.com/void-linux/void-packages/issues/7195#issuecomment-454897563

Tests are failing on both 1.0 and 1.1rc3 and Manual (in markdown, which converts poorly to man in this specific instance) is missing on 1.0, also cross-compilation requires more manual work.

All this leads to GNU bc being just the more pragmatic option for us.

**Comment by gavinhoward** [comment]
Original author of the `bc` here.

I am about to do a release, but I can delay it to fix the problems outlined here.

For cross-compiling, see the [build manual](https://github.com/gavinhoward/bc/blob/master/manuals/build.md#cross-compiling), but I think the problem there is that you are not defining the environment variable `HOSTCC`. In order to have zero dependencies, my `bc` has to compile and run a small program to generate C strings from files. (This is to add a math library to the `bc`.) This program will be compiled by `HOSTCC`, which defaults to `CC` if it is not defined. To make cross compiling work, it should be as simple as:

```
CC=/path/to/cross/compiler HOSTCC=/path/to/native/compiler ./configure.sh -O3
make
```

As far as 1.1rc3 not passing the test suite, that particular message is because the `bc` that is being used to generate the tests does not like GNU print statement syntax; it is not my `bc` failing the test. To generate a Makefile without the generated tests, run:

```
CC=/path/to/native/compiler ./configure.sh -GO3
make
```

Then my `bc` should pass all of its tests.

As far as man pages go, I am working on the problem.

**Comment by maxice8** [comment]
> For cross-compiling, see the build manual, but I think the problem there is that you are not defining the environment variable HOSTCC. In order to have zero dependencies, my bc has to compile and run a small program to generate C strings from files. (This is to add a math library to the bc.) This program will be compiled by HOSTCC, which defaults to CC if it is not defined. To make cross compiling work, it should be as simple as:

I see, we define `HOST_CC` and `CC_FOR_BUILD` but not `HOSTCC`

but it also needs to use `HOSTCFLAGS`, `HOST_CFLAGS` or `CFLAGS_FOR_BUILD` 

All seems to be done except manpages.

**Comment by gavinhoward** [comment]
Okay. I will make my build system also recognize `HOST_CC` and `HOST_CFLAGS`.

As for man pages, I am in the middle of a reformatting to make my manuals work with `ronn`. Once I am done, I will commit a man page for each into the repo and hook them into the `make install` target. I should be done by early tomorrow.

**Comment by gavinhoward** [comment]
Okay, I have made a few changes.

`HOST_CC` and `HOST_CFLAGS` are now recognized in the build system (along with `HOSTCFLAGS`, but that doesn't matter).

The manpages now exist ([bc](https://github.com/gavinhoward/bc/blob/master/manuals/bc.1) and [dc](https://github.com/gavinhoward/bc/blob/master/manuals/dc.1)) and are hooked into the install system. They are installed into `$(DESTDIR)$(PREFIX)/$(MANPAGE_LOC)/man1`. I can change the name of `MANPAGE_LOC` to something that Void expects, if that would make things easier.

**Comment by gavinhoward** [comment]
Oh, and `MANPAGE_LOC` defaults to `share/man`.

**Comment by maxice8** [comment]
On Thu, Jan 17, 2019 at 10:22:29AM -0800, Gavin D. Howard wrote:
> Okay, I have made a few changes.
> 
> `HOST_CC` and `HOST_CFLAGS` are now recognized in the build system (along with `HOSTCFLAGS`, but that doesn't matter).
> 
> The manpages now exist ([bc](https://github.com/gavinhoward/bc/blob/master/manuals/bc.1) and [dc](https://github.com/gavinhoward/bc/blob/master/manuals/dc.1)) and are hooked into the install system. They are installed into `$(DESTDIR)$(PREFIX)/$(MANPAGE_LOC)/man1`. I can change the name of `MANPAGE_LOC` to something that Void expects, if that would make things easier.
> 
> -- 
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub:
> https://github.com/void-linux/void-packages/issues/7195#issuecomment-455275853

the most common variable for installing manpages is constructing MANDIR out of DESTDIR/PREFIX/DATADIR

DATADIR is /usr/share in most cases.

**Comment by gavinhoward** [comment]
Oh, thank you. Based on that information, I was finally able to find [where those variables are coming from](https://www.gnu.org/software/make/manual/html_node/Directory-Variables.html), and I have updated my build system to match.

It now accepts `DATAROOTDIR` (which defaults to `$PREFIX/share`), `DATADIR` (which defaults to `$DATAROOTDIR`), `MANDIR` (which defaults to `$DATADIR/man`), and `MAN1DIR` (which defaults to `$MANDIR/man1`).

I also added `BINDIR` (which defaults to `$PREFIX/bin`) for good measure.

`DESTDIR` is also taken into account.
